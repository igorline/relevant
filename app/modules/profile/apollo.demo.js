import React, { useEffect } from 'react';
import { View, BodyText } from 'modules/styled/uni';
import { useQuery } from '@apollo/react-hooks';
import gql from 'graphql-tag';

const SUBSCRIBE = gql`
  subscription {
    userUpdated {
      _id
      balance
      handle
      name
      email
    }
  }
`;

const GET_ME = gql`
  query {
    me(filter: { handle: "slava" }) {
      _id
      handle
      balance
      name
      email
    }
  }
`;

export default function Test() {
  const { data, loading, error, subscribeToMore } = useQuery(GET_ME);
  useEffect(() => {
    const unsubscribe = subscribeToMore({ document: SUBSCRIBE });
    return () => unsubscribe();
  }, [subscribeToMore]);
  if (loading) return <BodyText>Loading...</BodyText>;
  if (error) return <BodyText>ERROR: {error.message}</BodyText>;
  return (
    <View>
      <BodyText>
        handle: {data.me.handle}, email: {data.me.email}, balance={data.me.balance}
      </BodyText>
    </View>
  );
}
