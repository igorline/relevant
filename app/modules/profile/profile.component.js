import React, { Component } from 'react';
import PropTypes from 'prop-types';

import UAvatar from 'modules/user/UAvatar.component';
import { colors, sizing, mixins, fonts } from 'app/styles';
import CoinStat from 'modules/stats/coinStat.component';
import RStat from 'modules/stats/rStat.component';
import { View, Header, AltLink, SecondaryText, Text } from 'modules/styled/uni';
import ULink from 'modules/navigation/ULink.component';
import { css } from 'styled-components/primitives';
import Percent from 'modules/stats/percent.component';

const SettingsImage = require('app/public/img/settings.svg');
const InviteImage = require('app/public/img/invite.svg');

const linkStyle = css`
  ${fonts.altLink}
  ${mixins.color}
`;

export default class Profile extends Component {
  static propTypes = {
    actions: PropTypes.object,
    isOwner: PropTypes.bool,
    user: PropTypes.object
  };

  render() {
    const { user, isOwner, actions } = this.props;
    if (!user) {
      return <div className="profileContainer">User not found!</div>;
    }

    // TODO upload image button
    // let uploadImg;
    // if (this.props.auth.user && user._id === this.props.auth.user._id) {
    //   uploadImg = <button className={'uploadImg edit'}>Update Profile Image</button>;
    // }

    return (
      <View
        m={sizing(4)}
        display="flex"
        fdirection="row"
        align="flex-start"
        justify="space-between"
      >
        <UAvatar user={user} size={9} />
        <View ml={sizing(2)} grow={1}>
          <View fdirection="row" display="flex" justify="space-between" align="baseline">
            <View display="flex" fdirection="row" align="baseline">
              <Header mr={sizing(2)}>{user.name}</Header>
              <View display="flex" fdirection="row" align={'baseline'}>
                <RStat size={1.75} user={user} mr={2} align="baseline" />
                <Percent size={1.75} user={user} align="baseline" />
                <CoinStat size={1.75} user={user} isOwner={isOwner} align="baseline" />
              </View>
            </View>
            {isOwner ? (
              <View align={'center'}>
                <ULink
                  onClick={() => {
                    actions.logoutAction(user);
                  }}
                  onPress={() => {
                    actions.logoutAction(user);
                  }}
                  color={colors.blue}
                  to="#"
                >
                  Logout
                </ULink>
              </View>
            ) : (
              <View />
            )}
          </View>
          {user.bio ? (
            <View mt={3}>
              <SecondaryText>{user.bio}</SecondaryText>
            </View>
          ) : null}
          {isOwner ? (
            <View fdirection="row" align="center">
              <AltLink ml={sizing(0.5)} mr={sizing(0.5)}>
                <ULink
                  c={colors.black}
                  to="/settings"
                  hc={colors.secondaryText}
                  styles={linkStyle}
                  onPress={e => {
                    e.preventDefault();
                    actions.showModal('settings');
                  }}
                  onClick={e => {
                    e.preventDefault();
                    actions.showModal('settings');
                  }}
                >
                  <Text fdirection="row" align="center">
                    <View mr={sizing(0.5)}>
                      <SettingsImage h={sizing(2)} w={sizing(2)} bg={colors.grey} />
                    </View>
                    Settings
                  </Text>
                </ULink>
              </AltLink>

              <AltLink ml={sizing(0.5)} mr={sizing(0.5)}>
                <ULink
                  c={colors.black}
                  to="/invites"
                  hc={colors.secondaryText}
                  styles={linkStyle}
                  onPress={e => {
                    e.preventDefault();
                    actions.showModal('invite');
                  }}
                  onClick={e => {
                    e.preventDefault();
                    actions.showModal('invite');
                  }}
                >
                  <Text fdirection="row" align="center" ml={sizing(1)}>
                    <View mr={sizing(0.5)}>
                      <InviteImage h={sizing(2)} w={sizing(2)} bg={colors.grey} />
                    </View>
                    Invite Friend
                  </Text>
                </ULink>
              </AltLink>
            </View>
          ) : null}
        </View>
      </View>
    );
  }
}
