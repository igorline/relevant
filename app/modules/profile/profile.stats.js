import React from 'react';
import PropTypes from 'prop-types';
import CoinStat from 'modules/stats/coinStat.component';
import RStat from 'modules/stats/rStat.component';
import Percent from 'modules/stats/percent.component';
import { Text } from 'modules/styled/uni';

ProfileStats.propTypes = {
  user: PropTypes.object,
  isOwner: PropTypes.bool
};

export default function ProfileStats({ user, isOwner }) {
  return (
    <Text>
      <RStat pr={1.5} inline={1} size={1.75} user={user} align="baseline" />
      {'  '}
      <Percent size={1.75} user={user} align="baseline" />
      {'   '}
      <CoinStat inline={1} size={1.75} user={user} isOwner={isOwner} align="baseline" />
    </Text>
  );
}
