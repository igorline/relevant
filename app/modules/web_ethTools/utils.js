import Web3 from 'web3';
import { INFURA_PROTOCOL, INFURA_NETWORK, INFURA_API_KEY } from 'core/config';

let web3;
let rpcUrl;

const defaultOptions = { _rpcUrl: getRpcUrl(), metamask: null };

export function getMetamask() {
  return typeof window !== 'undefined' ? window.ethereum : null;
}

export function getProvider(options = defaultOptions) {
  return options.metamask || options._rpcUrl !== defaultOptions._rpcUrl
    ? initProvider(options)
    : web3 || initProvider();
}

export function initProvider(options = defaultOptions) {
  const provider = createProvider(options);
  web3 = new Web3(provider);
  return web3;
}

export function getBN(value, web3Instance = web3) {
  return new web3Instance.utils.BN(value);
}

export function getRpcUrl() {
  return rpcUrl || buildRpcUrl();
}

export function createProvider(options = defaultOptions) {
  if (options.metamask) return options.metamask;
  return options._rpcUrl.slice(0, 2) === 'ws'
    ? new Web3.providers.WebsocketProvider(options._rpcUrl)
    : new Web3.providers.HttpProvider(options._rpcUrl);
}

export function buildRpcUrl(
  protocol = INFURA_PROTOCOL,
  network = INFURA_NETWORK,
  apiKey = INFURA_API_KEY
) {
  rpcUrl = `${protocol}s://${network}.infura.io/${
    protocol === 'ws' ? 'ws/' : ''
  }v3/${apiKey}`;
  return rpcUrl;
}
