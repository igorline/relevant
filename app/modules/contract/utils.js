import pickBy from 'lodash.pickby';
import { types } from 'core/contracts';
import { getProvider, getBN } from 'modules/web_ethTools/utils';

export function parseBN(value, web3Instance = getProvider()) {
  return value && value.get ? getBN(value.get('_hex'), web3Instance) : value;
}

// TODO -- provide better method sorting from statesauce
export const readableMethods = Object.keys(
  pickBy(
    types.methods,
    (_, method) =>
      !types.methods[method].send &&
      method !== 'balanceOf' &&
      method !== 'isMinter' &&
      method !== 'allowance' &&
      method !== 'partialSum' &&
      method !== 'nonceOf' &&
      method !== 'isOwner' &&
      method !== 'roundDecay'
  )
);
