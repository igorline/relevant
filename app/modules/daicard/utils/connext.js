/* eslint no-console: 0 */
import * as Connext from 'connext';
import interval from 'interval-promise';
import * as eth from 'ethers';
import { localStorage } from 'app/utils/storage';

const tokenAbi = require('./../abi/humanToken.json');

const { Big, minBN } = Connext.big;
const { CurrencyType, CurrencyConvertable } = Connext.types;
const { getExchangeRates, hasPendingOps } = new Connext.Utils();

const hubUrlRinkeby = 'https://rinkeby.hub.connext.network/api/hub';
const hubUrlMainnet = 'https://hub.connext.network/api/hub';

// Constants for channel max/min - this is also enforced on the hub
export const DEPOSIT_ESTIMATED_GAS = Big('700000'); // 700k gas
export const HUB_EXCHANGE_CEILING = eth.constants.WeiPerEther.mul(Big(69)); // 69 TST
export const CHANNEL_DEPOSIT_MAX = eth.constants.WeiPerEther.mul(Big(30)); // 30 TST
export const MAX_GAS_PRICE = Big('20000000000'); // 20 gWei

const RPC = 'RINKEBY';

let state = {
  isInitialized: null,
  connext: null,
  browserMinimumBalance: null,
  customWeb3: null,
  tokenContract: null,
  address: null,
  hubUrl: null
};

let updateReduxState;

export async function initConnext({ updateState, reinit }) {
  try {
    if (state.isInitialized && !reinit) return state.connext;
    updateReduxState = updateState;
    setState({ isInitialized: false });

    let mnemonic = localStorage.getItem('mnemonic');
    if (!mnemonic) {
      mnemonic = eth.Wallet.createRandom().mnemonic;
      localStorage.setItem('mnemonic', mnemonic);
    }

    await setConnext({ mnemonic });
    await setTokenContract();
    addConnextHandler();
    poller();

    setState({ isInitialized: true });

    console.log(state);
    return state.connext;
  } catch (err) {
    return null;
    console.log(err); // eslint-disable-line
  }
}

function setState(newState) {
  state = {
    ...state,
    ...newState
  };
  updateReduxState(newState);
}

async function setConnext({ rpc, mnemonic }) {
  let ethprovider;
  let hubUrl;

  switch (RPC) {
    // case 'LOCALHOST':
    //   ethprovider = localProvider;
    //   hubUrl = hubUrlLocal;
    //   break;
    case 'RINKEBY':
      ethprovider = new eth.getDefaultProvider('rinkeby'); // eslint-disable-line
      hubUrl = hubUrlRinkeby;
      break;
    case 'MAINNET':
      hubUrl = hubUrlMainnet;
      ethprovider = new eth.getDefaultProvider(); // eslint-disable-line
      break;
    default:
      throw new Error(`Unrecognized rpc: ${rpc}`);
  }

  const opts = {
    hubUrl,
    mnemonic
  };

  // *** Instantiate the connext client ***
  const connext = await Connext.getConnextClient(opts);

  const address = await connext.wallet.getAddress();

  console.log('Successfully set up connext! Connext config:');
  console.log(`  - tokenAddress: ${connext.opts.tokenAddress}`);
  console.log(`  - hubAddress: ${connext.opts.hubAddress}`);
  console.log(`  - contractAddress: ${connext.opts.contractAddress}`);
  console.log(`  - ethNetworkId: ${connext.opts.ethNetworkId}`);

  return setState({
    connext,
    tokenAddress: connext.opts.tokenAddress,
    channelManagerAddress: connext.opts.contractAddress,
    hubWalletAddress: connext.opts.hubAddress,
    ethNetworkId: connext.opts.ethNetworkId,
    address,
    ethprovider
  });
}

function addConnextHandler() {
  const { connext } = state;
  let { browserMinimumBalance } = state;
  connext.on('onStateChange', async connextState => {
    try {
      console.log('Connext state changed:', connextState);
      setState({
        connextState,
        channelState: connextState.persistent.channel,
        runtime: connextState.runtime,
        exchangeRate: connextState.runtime.exchangeRate
          ? connextState.runtime.exchangeRate.rates.USD
          : 0
      });
      if (!browserMinimumBalance) {
        browserMinimumBalance = await setBrowserWalletMinimumBalance();
        setState({ browserMinimumBalance });
      }
      // checkStatus();
    } catch (err) {
      console.log(err);
    }
  });
}

async function setBrowserWalletMinimumBalance() {
  const { connextState, ethprovider } = state;
  const gasEstimateJson = await eth.utils.fetchJson({
    url: 'https://ethgasstation.info/json/ethgasAPI.json'
  });
  const providerGasPrice = await ethprovider.getGasPrice();
  // multiply gas price by two to be safe
  let currentGasPrice = Math.round((gasEstimateJson.average / 10) * 2);
  // dont let gas price be any higher than the max
  currentGasPrice = eth.utils.parseUnits(
    minBN(Big(currentGasPrice.toString()), MAX_GAS_PRICE).toString(),
    'gwei'
  );
  // unless it really needs to be: average eth gas station price w ethprovider's
  currentGasPrice = currentGasPrice.add(providerGasPrice).div(eth.constants.Two);
  console.log(`Gas Price = ${currentGasPrice}`);

  // default connext multiple is 1.5, leave 2x for safety
  const totalDepositGasWei = DEPOSIT_ESTIMATED_GAS.mul(Big(2)).mul(currentGasPrice);

  // add dai conversion
  const minConvertable = new CurrencyConvertable(
    CurrencyType.WEI,
    totalDepositGasWei,
    () => getExchangeRates(connextState)
  );
  const browserMinimumBalance = {
    wei: minConvertable.toWEI().amount,
    dai: minConvertable.toUSD().amount
  };
  setState({ browserMinimumBalance });
  return browserMinimumBalance;
}

async function setTokenContract() {
  try {
    const { tokenAddress, ethprovider } = state;
    const tokenContract = new eth.Contract(tokenAddress, tokenAbi, ethprovider);
    return setState({ tokenContract });
  } catch (e) {
    console.log('Error setting token contract');
    return console.log(e);
  }
}

async function poller() {
  await autoDeposit();
  await autoSwap();
  interval(async () => autoDeposit(), 5000);
  interval(async () => autoSwap(), 1000);
}

async function autoDeposit() {
  const {
    address,
    tokenContract,
    connextState,
    tokenAddress,
    connext,
    browserMinimumBalance,
    ethprovider
  } = state;

  if (!connext || !browserMinimumBalance) return;

  const balance = await ethprovider.getBalance(address);

  let tokenBalance = '0';
  try {
    tokenBalance = await tokenContract.balanceOf(address);
  } catch (e) {
    console.warn(
      `Error fetching token balance, are you sure the token address (addr: ${tokenAddress}) is correct for the selected network (id: ${JSON.stringify(
        await ethprovider.getNetwork()
      )}))? Error: ${e.message}`
    );
    return;
  }

  if (balance.gt('0') || tokenBalance.gt('0')) {
    const minWei = Big(browserMinimumBalance.wei);
    if (Big(balance).lt(minWei)) {
      // don't autodeposit anything under the threshold
      // update the refunding variable before returning
      // We hit this repeatedly after first deposit & we have dust left over
      // No need to clutter logs w the below
      // console.log(`Current balance is ${balance.toString()},
      // less than minBalance of ${minWei.toString()}`);
      return;
    }
    // only proceed with deposit request if you can deposit
    if (!connextState) {
      return;
    }
    if (
      // something was submitted
      connextState.runtime.deposit.submitted ||
      connextState.runtime.withdrawal.submitted ||
      connextState.runtime.collateral.submitted
    ) {
      console.log('Deposit or withdrawal transaction in progress, will not auto-deposit');
      return;
    }

    const channelDeposit = {
      amountWei: Big(balance).sub(minWei),
      amountToken: tokenBalance
    };

    if (channelDeposit.amountWei === '0' && channelDeposit.amountToken === '0') {
      return;
    }

    await state.connext.deposit({ amountWei: channelDeposit.amountWei.toString() });
  }
}

async function autoSwap() {
  const { channelState, connextState } = state;
  if (!connextState || hasPendingOps(channelState)) {
    return;
  }
  const weiBalance = Big(channelState.balanceWeiUser);
  const tokenBalance = Big(channelState.balanceTokenUser);
  if (channelState && weiBalance.gt(Big('0')) && tokenBalance.lte(HUB_EXCHANGE_CEILING)) {
    await state.connext.exchange(channelState.balanceWeiUser, 'wei');
  }
}
