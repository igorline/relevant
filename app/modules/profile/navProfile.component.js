import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import UAvatar from 'modules/user/UAvatar.component';
import CoinStat from 'modules/stats/coinStat.component';
import RStat from 'modules/stats/rStat.component';
import ULink from 'modules/navigation/ULink.component';
import * as navigationActions from 'modules/navigation/navigation.actions';
import ReactTooltip from 'react-tooltip';

import { sizing, colors } from 'app/styles';
import styled from 'styled-components/primitives';
import { Header, View, SecondaryText, CTALink, Text } from 'modules/styled/uni';
import { computeUserPayout } from 'app/utils/rewards';

const WalletInfo = styled.View`
  display: flex;
  flex-direction: column;
  padding-left: ${sizing(2)};
  flex-shrink: 1;
`;

export class NavProfile extends Component {
  static propTypes = {
    user: PropTypes.object,
    earnings: PropTypes.object,
    mobile: PropTypes.bool,
    actions: PropTypes.object
  };

  componentDidMount() {
    if (ReactTooltip.rebuild) ReactTooltip.rebuild();
  }

  render() {
    const { user, earnings, mobile, actions } = this.props;
    if (!user) return null;

    // TODO optimize this so its not on every render?
    let pendingPayouts = 0;
    earnings.pending.forEach(id => {
      const earning = earnings.entities[id];
      pendingPayouts += computeUserPayout(earning);
    });

    const p = mobile ? 2 : 4;
    const hideGetTokens = user.twitterId && user.confirmed;

    return (
      <View bb flex={1}>
        <View p={p} pb={p + 1}>
          <View fdirection={'row'} justify="space-between" align="center">
            <Header>{user.name}</Header>
            <ULink hu to="/user/wallet" onPress={() => actions.goToTab('wallet')}>
              <CTALink c={colors.blue}>My Wallet</CTALink>
            </ULink>
          </View>

          <View fdirection={'row'} align={'center'} mt={4}>
            <UAvatar
              user={user}
              size={8}
              noName
              goToProfile={() => actions.goToTab('myProfile')}
            />
            <WalletInfo>
              <View fdirection={'row'}>
                <ULink to="/user/wallet" onPress={() => actions.push('statsView')}>
                  <RStat
                    user={user}
                    align="center"
                    data-for="mainTooltip"
                    data-tip={JSON.stringify({
                      type: 'TEXT',
                      props: {
                        text:
                          'Earn Reputation by posting comments.\nThe higher your score, the more weight your votes have.'
                      }
                    })}
                  />
                </ULink>
                <ULink to="/user/wallet" onPress={() => actions.goToTab('wallet')}>
                  <CoinStat
                    user={user}
                    isOwner={true}
                    align="center"
                    data-for="mainTooltip"
                    data-tip={JSON.stringify({
                      type: 'TEXT',
                      props: {
                        text:
                          'Get coins by upvoting quality links.\nThe higher your Reputation the more coins you earn.'
                      }
                    })}
                  />
                </ULink>
              </View>
              <View fdirection={'row'} align={'baseline'} color={colors.grey} mt={2}>
                <SecondaryText fs={1.5}>Pending Rewards: </SecondaryText>
                <CoinStat
                  size={1.5}
                  mr={1.5}
                  fs={1.5}
                  secondary
                  c={colors.black}
                  amount={pendingPayouts}
                  align={'baseline'}
                  data-for="mainTooltip"
                  data-tip={JSON.stringify({
                    type: 'TEXT',
                    props: {
                      text:
                        'These are your projected earnings for upvoting quality posts.\nRewards are paid out 3 days after a link is posted.'
                    }
                  })}
                />
              </View>
            </WalletInfo>
          </View>

          <View fdirection={'row'} align={'baseline'} mt={3}>
            {hideGetTokens ? null : (
              <ULink
                to="/user/wallet"
                c={colors.blue}
                hu
                onPress={e => {
                  e.preventDefault();
                  actions.showModal('getTokens');
                }}
                onClick={e => {
                  e.preventDefault();
                  actions.showModal('getTokens');
                }}
              >
                <CTALink c={colors.blue}>Get Tokens</CTALink>
              </ULink>
            )}
            {hideGetTokens ? null : <Text> &nbsp;&nbsp; </Text>}
            <ULink
              to="/user/wallet"
              ml={1}
              c={colors.blue}
              hu
              onPress={e => {
                e.preventDefault();
                actions.showModal('invite');
              }}
              onClick={e => {
                e.preventDefault();
                actions.showModal('invite');
              }}
            >
              <CTALink c={colors.blue}>Invite Friends</CTALink>
            </ULink>
          </View>
        </View>
      </View>
    );
  }
}

const mapStateToProps = state => ({
  user: state.auth.user,
  earnings: state.earnings
});

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators(
    {
      ...navigationActions
    },
    dispatch
  )
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(NavProfile);
