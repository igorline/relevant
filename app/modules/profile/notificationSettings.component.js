import React, { Component } from 'react';
import PropTypes from 'prop-types';
import get from 'lodash.get';
import { View, Header, SecondaryText, Title, BodyText } from 'modules/styled/uni';
import { connect } from 'react-redux';
import {
  updateNotificationSettings,
  enableMobileNotifications
} from 'modules/auth/auth.actions';
import { enableDesktopNotifications } from 'modules/activity/activity.actions';
import { bindActionCreators } from 'redux';
import ToggleSwitch from 'modules/ui/toggleswitch.component';
import { colors } from 'app/styles';
import { capitalize } from 'utils/numbers';

const NOTIFICATION_DETAILS = {
  email: {
    personal: {
      label: 'Replies, mentions and earnings',
      description:
        'Get notified when you earn rewards, someone replies to your comments or mentions you in a post.'
    },
    general: {
      label: 'General notifications',
      description:
        'Get notified about new posts and comments on posts you have interacted with.'
    },
    digest: {
      label: 'Email digest',
      description:
        'Receive periodic emails with the top posts from your favorite communities.'
    }
  },
  mobile: {
    all: {
      label: 'Mobile notifications',
      description: 'Receive mobile app notifications for any community activity.'
    }
  },
  desktop: {
    all: {
      label: 'Desktop notifications',
      description: 'Receive notifications in a browser for any community activity.'
    }
  }
};

const NotificationToggle = ({ parent, label, notification, onChange }) => {
  const details = NOTIFICATION_DETAILS[parent][label];
  if (!details) return null;
  return (
    <View fdirection="row" mt={3}>
      <View fdirection="row" align="center" flex={1}>
        <ToggleSwitch
          isOn={!!notification}
          onColor={colors.green}
          offColor={colors.grey}
          size="custom"
          onToggle={isOn => {
            const body = {};
            body[parent] = {};
            body[parent][label] = isOn;
            onChange(body);
          }}
        />
        <View fdirection="column" ml={1.5} shrink={1} wrap>
          {details.label ? <Title>{details.label}</Title> : null}
          {details.description ? <BodyText>{details.description}</BodyText> : null}
        </View>
      </View>
    </View>
  );
};

NotificationToggle.propTypes = {
  parent: PropTypes.string,
  label: PropTypes.string,
  notification: PropTypes.bool,
  onChange: PropTypes.func
};

const NotificationSet = ({ label, notifications, onChange }) => {
  const sortedNotifications = Object.keys(notifications).sort();
  return (
    <View mt={4}>
      <SecondaryText>{capitalize(label)} Notifications</SecondaryText>
      {notifications &&
        sortedNotifications.map(k => (
          <NotificationToggle
            key={k}
            label={k}
            parent={label}
            notification={notifications && notifications[k]}
            onChange={onChange}
          />
        ))}
    </View>
  );
};

NotificationSet.propTypes = {
  label: PropTypes.string,
  notifications: PropTypes.object,
  onChange: PropTypes.func
};

class NotificationSettings extends Component {
  async componentDidUpdate(prevProps) {
    const { actions, user } = this.props;
    const { notificationSettings } = user;
    if (
      !get(prevProps, 'user.notificationSettings.desktop.all') &&
      get(notificationSettings, 'desktop.all')
    ) {
      // TODO fallback for when notifications were previously disabled
      actions.enableDesktopNotifications();
    }

    if (
      !get(prevProps, 'user.notificationSettings.mobile.all') &&
      get(notificationSettings, 'mobile.all')
    ) {
      // TODO fallback for when notifications were previously disabled
      actions.enableMobileNotifications(user);
    }
    return null;
  }

  render() {
    const { user, actions } = this.props;
    if (!user) {
      return null;
    }
    const { notificationSettings } = user;
    if (!notificationSettings) return null;
    const sortedSettings = Object.keys(notificationSettings).sort();
    return (
      <View display="flex" fdirection="column">
        <Header>Notifications</Header>
        {sortedSettings.map(key => (
          <NotificationSet
            key={key}
            label={key}
            notifications={notificationSettings[key]}
            onChange={actions.updateNotificationSettings}
          />
        ))}
      </View>
    );
  }
}

NotificationSettings.propTypes = {
  user: PropTypes.object,
  actions: PropTypes.object
};

const mapStateToProps = state => ({
  user: state.auth.user,
  enableReinitialize: true
});

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators(
    {
      updateNotificationSettings,
      enableDesktopNotifications,
      enableMobileNotifications
    },
    dispatch
  )
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(NotificationSettings);
