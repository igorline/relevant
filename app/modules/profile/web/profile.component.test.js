import React from 'react';
import renderer from 'react-test-renderer';
import { Profile } from 'modules/profile/web/profile.component';
import 'jest-styled-components';
import { MemoryRouter } from 'react-router-dom';
import { auth, user1, usersState } from 'app/utils/testData';

const props = {
  actions: { logout: jest.fn() },
  usersState,
  auth,
  match: { params: { id: user1.handle } }
};

test('Snapshot Profile Web', () => {
  const tree = renderer.create(
    <MemoryRouter><Profile {...props} /></MemoryRouter>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
