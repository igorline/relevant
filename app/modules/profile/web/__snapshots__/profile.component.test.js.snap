// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Post Web 1`] = `
.c0 {
  display: inline-block;
  width: 100%;
  text-align: right;
  padding-right: 2em;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0 color:function (props) {
  return: props.color || 'black';
}

<div
  className="profileContainer"
>
  <div
    className="profileHero"
  >
    <a
      className="c0"
      color="#4d4eff"
      onClick={[Function]}
    >
       Logout
    </a>
    <a
      className="avatar"
      href="/user/profile/handle1"
      onClick={[Function]}
      style={
        Object {
          "backgroundImage": "url(userImage1.jpg)",
          "height": undefined,
          "width": undefined,
        }
      }
    >
      Name1
    </a>
    <div
      className="name"
    >
      Name1
    </div>
    <div
      className="relevance"
    >
      <img
        alt="Relevance"
        className="r"
        src="/img/r-emoji.png"
      />
      11
      <img
        alt="Coins"
        className="coin"
        src="/img/relevantcoin.png"
      />
      111
    </div>
    <div
      className="subscribers"
    >
      Subscribers: 
      <b />
       â€¢ 
      Subscribed to: 
      <b />
    </div>
    <div
      className="tags"
    >
      Expertise: 
    </div>
  </div>
</div>
`;
